browser/ui/email_aliases/email_aliases_controller.cc

```diff
@@ -5,24 +5,136 @@
 #include "content/public/browser/context_menu_params.h"
 #include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_ui.h"
 #include "third_party/blink/public/mojom/forms/form_control_type.mojom-shared.h"
+#include "ui/web_dialogs/web_dialog_delegate.h"
 #include "url/gurl.h"
 
 namespace {
 constexpr char kEmailAliasesSettingsURL[] = "brave://settings/email-aliases";
+constexpr int kDialogWidth = 420;
+constexpr gfx::Size kDialogMinSize(kDialogWidth, 336);
+constexpr gfx::Size kDialogMaxSize(kDialogWidth, 794);
 
 bool g_autoclose_bubble_for_testing = false;
 
 }  // namespace
 
 namespace email_aliases {
 
+namespace {
+
+class EmailAliasesDialogDelegate
+    : public ui::WebDialogDelegate,
+      email_aliases::mojom::EmailAliasesPanelHandler,
+      content::WebContentsObserver {
+ public:
+  ~EmailAliasesDialogDelegate() override = default;
+
+  static ConstrainedWebDialogDelegate* Show(
+      EmailAliasesController& controller,
+      content::WebContents* initiator,
+      content::GlobalRenderFrameHostId field_render_frame_host_id,
+      uint64_t field_renderer_id) {
+    return ShowConstrainedWebDialogWithAutoResize(
+        initiator->GetBrowserContext(),
+        base::WrapUnique(new EmailAliasesDialogDelegate(
+            controller, field_render_frame_host_id, field_renderer_id)),
+        initiator, kDialogMinSize, kDialogMaxSize);
+  }
+
+ private:
+  EmailAliasesDialogDelegate(
+      EmailAliasesController& controller,
+      content::GlobalRenderFrameHostId field_render_frame_host_id,
+      uint64_t field_renderer_id)
+      : email_aliases_controller_(controller),
+        field_render_frame_host_id_(field_render_frame_host_id),
+        field_renderer_id_(field_renderer_id) {
+    set_delete_on_close(false);
+    set_dialog_content_url(GURL(kEmailAliasesPanelURL));
+    set_show_dialog_title(false);
+    set_close_dialog_on_escape(true);
+    set_can_close(true);
+  }
+
+  // ui::WebDialogDelegate:
+  void OnDialogShown(content::WebUI* webui) override {
+    if (auto* ui = webui->GetController()->GetAs<EmailAliasesPanelUI>()) {
+      ui->SetHandlerDelegate(this);
+      Observe(webui->GetWebContents());
+    }
+  }
+
+  // email_aliases::mojom::EmailAliasesPanelHandler:
+  void OnAliasCreated(const std::string& email) override {
+    Close();
+
+    if (email.empty()) {
+      return;
+    }
+    auto* field_render_frame =
+        content::RenderFrameHost::FromID(field_render_frame_host_id_);
+    if (!field_render_frame) {
+      return;
+    }
+    auto* autofill_driver =
+        autofill::ContentAutofillDriver::GetForRenderFrameHost(
+            field_render_frame);
+    if (!autofill_driver) {
+      return;
+    }
+    autofill_driver->GetAutofillAgent()->ApplyFieldAction(
+        autofill::mojom::FieldActionType::kReplaceAll,
+        autofill::mojom::ActionPersistence::kFill,
+        autofill::FieldRendererId(field_renderer_id_),
+        base::UTF8ToUTF16(email));
```

Comment:
this is kind interesting - why don't we just let the UI know via mojom what the new email is?

It seems a bit brittle using autofill for it.